### PR TITLE
get_kubeconfig.sh: make output file path configurable

### PR DESCRIPTION
**Pull request by hardys**
Currently we leak files into /tmp which isn't ideal in many cases, so add the option to specify a predictable filename via the env file

### CODE REVIEW

**Comment by hardys** [comment]
> Would you mind adding the parameter to the readme as well?

Good point, done! :)

**Comment by e-minguez** [comment]
> > Would you mind adding the parameter to the readme as well?
> 
> Good point, done! :)

sorry, I was talking about https://github.com/suse-edge/misc/blob/c8406c949cf7b32ac02d65859e41a31779596ee5/slemicro/README.md?plain=1#L129 😅

**Comment by hardys** [comment]
> > > Would you mind adding the parameter to the readme as well?
> > 
> > 
> > Good point, done! :)
> 
> sorry, I was talking about
> 
> https://github.com/suse-edge/misc/blob/c8406c949cf7b32ac02d65859e41a31779596ee5/slemicro/README.md?plain=1#L129
> 
> 😅

Oh I see, I missed that we had that in the README as well as the `env.example` 😄 will add!

I guess another approach instead of this override would be to drop the mktemp completely and have the kubeconfig default to `${BASEDIR}/vm-${VMNAME}.kubeconfig` ?

**Comment by e-minguez** [comment]
> > > > Would you mind adding the parameter to the readme as well?
> > > 
> > > 
> > > Good point, done! :)
> > 
> > 
> > sorry, I was talking about
> > https://github.com/suse-edge/misc/blob/c8406c949cf7b32ac02d65859e41a31779596ee5/slemicro/README.md?plain=1#L129
> > 
> > 😅
> 
> Oh I see, I missed that we had that in the README as well as the `env.example` 😄 will add!
> 
> I guess another approach instead of this override would be to drop the mktemp completely and have the kubeconfig default to `${BASEDIR}/vm-${VMNAME}.kubeconfig` ?

I've been thinking about this and what about:

* if the script was executed with `-f`, then create the kubeconfig just like the parameters file but with the .kubeconfig extension
* otherwise as you said, `${BASEDIR}/vm-${VMNAME}.kubeconfig`

OTOH not sure if this is going to complicate things for little gains or just keep it as your first proposal (mktemp if not provided)

**Comment by hardys** [comment]
> > > > > Would you mind adding the parameter to the readme as well?
> > > > 
> > > > 
> > > > Good point, done! :)
> > > 
> > > 
> > > sorry, I was talking about
> > > https://github.com/suse-edge/misc/blob/c8406c949cf7b32ac02d65859e41a31779596ee5/slemicro/README.md?plain=1#L129
> > > 
> > > 😅
> > 
> > 
> > Oh I see, I missed that we had that in the README as well as the `env.example` 😄 will add!
> > I guess another approach instead of this override would be to drop the mktemp completely and have the kubeconfig default to `${BASEDIR}/vm-${VMNAME}.kubeconfig` ?
> 
> I've been thinking about this and what about:
> 
> * if the script was executed with `-f`, then create the kubeconfig just like the parameters file but with the .kubeconfig extension
> * otherwise as you said, `${BASEDIR}/vm-${VMNAME}.kubeconfig`
> 
> OTOH not sure if this is going to complicate things for little gains or just keep it as your first proposal (mktemp if not provided)

That sounds reasonable to me - honestly I don't have a strong preference, I can easily rework this as you describe if you prefer that approach 👍

**Comment by e-minguez** [comment]
> > > > > > Would you mind adding the parameter to the readme as well?
> > > > > 
> > > > > 
> > > > > Good point, done! :)
> > > > 
> > > > 
> > > > sorry, I was talking about
> > > > https://github.com/suse-edge/misc/blob/c8406c949cf7b32ac02d65859e41a31779596ee5/slemicro/README.md?plain=1#L129
> > > > 
> > > > 😅
> > > 
> > > 
> > > Oh I see, I missed that we had that in the README as well as the `env.example` 😄 will add!
> > > I guess another approach instead of this override would be to drop the mktemp completely and have the kubeconfig default to `${BASEDIR}/vm-${VMNAME}.kubeconfig` ?
> > 
> > 
> > I've been thinking about this and what about:
> > 
> > * if the script was executed with `-f`, then create the kubeconfig just like the parameters file but with the .kubeconfig extension
> > * otherwise as you said, `${BASEDIR}/vm-${VMNAME}.kubeconfig`
> > 
> > OTOH not sure if this is going to complicate things for little gains or just keep it as your first proposal (mktemp if not provided)
> 
> That sounds reasonable to me - honestly I don't have a strong preference, I can easily rework this as you describe if you prefer that approach 👍

Sorry, I think the best way would be a third way 😅 , the `curl` or `grub2-mkconfig` way:

* The kubeconfig is show in the stdout by default
* If using a flag (`-o` to be compliant with both `curl` & `grub2-mkconfig`), it goes to that destination.

I think this simplifies the whole logic and it is more flexible, like:

```
# classic example if using just the ~/.kube/config file
./get-kubeconfig.sh -f foobar && mv foobar.kubeconfig ~/.kube/config
# vs
./get-kubeconfig -f foobar -o ~/.kube/config
# or
./get-kubeconfig -f foobar > ~/.kube/config
```

In any case... whatever you choose is better than my poor attempt using `mkdir` so I'll approve any approach you decide :D

**Comment by hardys** [comment]
>  ./get-kubeconfig.sh -f foobar && mv foobar.kubeconfig ~/.kube/config

I'm fine with adding the `-o` but I'd prefer not to rely on the `-f` filename to name the kubeconfig, because that won't work in the default case (read from `.env` or use script defaults) - if we use a name derived from `${VMNAME}` that will work regardless of the mode of invocation and will be more consistent?

**Comment by e-minguez** [comment]
> > ./get-kubeconfig.sh -f foobar && mv foobar.kubeconfig ~/.kube/config
> 
> I'm fine with adding the `-o` but I'd prefer not to rely on the `-f` filename to name the kubeconfig, because that won't work in the default case (read from `.env` or use script defaults) - if we use a name derived from `${VMNAME}` that will work regardless of the mode of invocation and will be more consistent?

sorry, I explained myself very poorly 😓 that first line was just an example if we implement the 'get the name from the filename or the name' vs the other two lines are examples on how to do it with the `-o` flag.

File: slemicro/README.md [hunk 1]
```diff
@@ -261,19 +261,22 @@ The previous environment variables can be used but it requires a few less.
...
-Usage: ./get_kubeconfig.sh [-f <path/to/variables/file>] [-n <vmname>] [-i <vmip>]
+Usage: ./get_kubeconfig.sh [-f <path/to/variables/file>] [-n <vmname>] [-i <vmip>] [-o <filename>]
 ```
 
 You can use the script in combination with the `create_vm.sh` one as:
 
 ```bash
 $ ./create_vm.sh -f vm-foobar
-$ KUBECONFIG=$(./get_kubeconfig.sh -w -f vm-foobar)
+$ ./get_kubeconfig.sh -w -f vm-foobar -o $KUBECONFIG
```
Comment (hardys) [inline slemicro/README.md:279]: @e-minguez is it a problem that this existing usage pattern will differ?

For me I like to set a consistent KUBECONFIG in my shell environment, then each time I re-run `create_vm.sh` I overwrite the file, which we can still do but without leaking the files to /tmp?

**Comment by hardys** [comment]
@e-minguez sorry I forgot about this PR - now updated to align with the suggestions to align with curl-ish behaviour

File: slemicro/README.md (hunk 1 above)
Comment (e-minguez) [reply slemicro/README.md:279]: It is ok to me!

File: slemicro/get_kubeconfig.sh [hunk 2]
```diff
@@ -118,6 +123,10 @@ else
 	curl -sk "https://${RANCHERHOSTNAME}/v3/clusters/local?action=generateKubeconfig" -X 'POST' -H 'content-type: application/json' -H "Authorization: Bearer ${TOKEN}" | jq -r .config > ${TMPKUBECONFIG} 
 fi
 
-echo ${TMPKUBECONFIG}
-
-exit 0
+if [ -z "${OUTPUT_FILE}" ]; then
+  cat echo ${TMPKUBECONFIG}
```
Comment (e-minguez) [inline slemicro/get_kubeconfig.sh:127]: Shouldn't this be just 'cat ${TMPKUBECONFIG}' ?

**Review by e-minguez — CHANGES_REQUESTED** [review]

File: slemicro/get_kubeconfig.sh (hunk 2 above)
Comment (hardys) [reply slemicro/get_kubeconfig.sh:127]: oops, yes! 😂

Comment (hardys) [reply slemicro/get_kubeconfig.sh]: Fixed

**Review by e-minguez — APPROVED** [review]

**Pull request merged**